kroxylicious-kms/src/main/java/io/kroxylicious/kms/service/KekId.java

```diff
@@ -0,0 +1,26 @@
+/*
+ * Copyright Kroxylicious Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package io.kroxylicious.kms.service;
+
+/**
+ * A semantic wrapper around KMS specific key id type
+ */
+public interface KekId {
+
+    /**
+     * Obtain the underlying key id
+     *
+     * @param <K> the type of the underlying Key
+     *
+     * @return the underlying key ID
+     */
+    <K> K getId(Class<K> keyType);
+
+    default boolean isUnencrypted() {
```

Comment:
This felt like it was going to a better place in the other PR, some kind of `EncryptionScheme` type object. It looks a bit clunky having `KekId` be something that is either `unencrypted` or supports an `id` being retrieved from it.

Reply:
Maybe another way to think about it, is the decision about whether it should be encrypted a KMS responsibility?

Reply:
Maybe the right answer is actually hybrid of the two but in both (and on main) it's the KMS making the decision.

The question is how to convey that.
1. A semantic null (as on main)
1. A sentinel ID (as in the other PR)
1. A property of the type (as here)
1. Some other idea I haven't had.


1. doesn't feel great, to me at least, and leads to implementation issues (e.g we can't use collectors.toMap, due to the null values)

With 2. It felt like we were making the same decision twice. Once in the KMS to catch the unknown alias exceptions and return and once outside to determine what was returned.

Reply:
Okay I see what you mean. With the templating approach we don't know up front which topic names will have KEKs and it's the existence of the key in the KMS that is deciding whether we encrypt.

But this logic is not something the KMS module is aware of or concerned with, is it? In main the decision to to convert unresolved alias exception to a meaningful null that means don't encrypt is a filter responsibility.

If this is going to remain a supported kek mapping technique it feels like an exception is wrong, since we would expect unresolved aliases under normal operation.

Reply:
so while it feels a bit like the same decision twice I think the `Kms` abstraction should be conveying whether it resolved and it's up to the encryption filter to decide what that means for encrypt. A lenient strategy might decide it means pass through the data unencrypted. A harsh strategy that says "all topics must have a key matching template X" would act differently.

Reply:
>If this is going to remain a supported kek mapping technique it feels like an exception is wrong, since we would expect unresolved aliases under normal operation.

I largely agree, I can see value in and uses for an exception but I do think failed mappings are relatively normal.  However I'm not sure how else to signal that it failed to resolve and keep the separation between the filter and KMS sides of the API.

>But this logic is not something the KMS module is aware of or concerned with, is it? In main the decision to to convert unresolved alias exception to a meaningful null that means don't encrypt is a filter responsibility.

Hmm not sure I agree. The conversion is happening in a `KekSelector` which is part of the "filter" not the KMS.

>A harsh strategy that says "all topics must have a kekId matching template X" would act differently.

Yeah but that strategy would just forward the exception in the form of a failed future from `selectKek`

Reply:
> However I'm not sure how else to signal that it failed to resolve

A result type class like `ResolvedKekId` or `UnresolvedKekId`, and then use exceptions for the exceptional cases.

> The conversion is happening in a KekSelector which is part of the "filter" not the KMS.
Yep, a filter responsibility. That's my point. A KekId (part of the KMS api) shouldn't decide which records should be encrypted or unencrypted. Maybe all I'm looking for is a rename of the method from `KekId#isUnencrypted()` to `KekId#isResolved()`

Reply:
though then it is kind of a bother that KekId may or may not really be an Id and it might be better name `KekIdResult` or similar.

Reply:
> A KekId (part of the KMS api) shouldn't decide which records should be encrypted or unencrypted.

But this PR doesn't do that, the KMS isn't making policy decisions. It asks the KMS for a value which represents un-encrypted things. The KMS still know nothing of records. However we still need to ask the KMS for what that ID is so we can have something which can be legitimately serialised. That is the KMS is responsible for defining the sentinel values based on which ever ID type it uses. I.e. a magic string or a the `NIL_UUID`

Reply:
>A result type class like ResolvedKekId or UnresolvedKekId, and then use exceptions for the exceptional cases.

Yeah Ok, I was being dense.

Reply:
>  However we still need to ask the KMS for what that ID is so we can have something which can be legitimately serialised

Ooh maybe I've missed something here. Are the nil/unencrypted kek ids serialized somewhere? I thought this was just a marker that would then be used to skip encryption.

Reply:
Hmm maybe I've pushed it too far as it should never actually make it to the wire. I do sort of think we should be able to treat the sentinel values as we would any other ID and thus they should be recognisable by the KMS and be capable of being written to the wire.

Reply:
This surfaced the need for https://github.com/kroxylicious/kroxylicious/issues/811